plugins/catalog-backend/src/tests/performance/getEntitiesPerformance.test.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { describePerformanceTest, performanceTraceEnabled } from './lib/env';
+import { startTestBackend, TestDatabases } from '@backstage/backend-test-utils';
+import { applyDatabaseMigrations } from '../../database/migrations';
+import {
+  SyntheticLoadEntitiesProcessor,
+  SyntheticLoadEntitiesProvider,
+  SyntheticLoadOptions,
+} from './lib/catalogModuleSyntheticLoadEntities';
+import { CatalogClient } from '@backstage/catalog-client';
+import {
+  coreServices,
+  createBackendModule,
+  createServiceFactory,
+} from '@backstage/backend-plugin-api';
+import { catalogProcessingExtensionPoint } from '@backstage/plugin-catalog-node/alpha';
+import { Knex } from 'knex';
+
+jest.setTimeout(600_000);
+
+const traceLog: typeof console.log = performanceTraceEnabled
+  ? console.log
+  : () => {};
+
+const completionPolling = async (load: SyntheticLoadOptions, knex: Knex) => {
+  const { baseEntitiesCount, childrenCount } = load;
+  const expectedTotal = baseEntitiesCount + baseEntitiesCount * childrenCount;
+
+  return new Promise<void>((resolve, reject) => {
+    const interval = setInterval(async () => {
+      try {
+        const stitchedCount = await knex('final_entities')
+          .count({ count: '*' })
+          .whereNotNull('final_entity')
+          .then(rows => Number(rows[0].count));
+
+        if (stitchedCount === expectedTotal) {
+          clearInterval(interval);
+          resolve();
+        }
+      } catch (error) {
+        clearInterval(interval);
+        reject(error);
+      }
+    }, 1000);
+  });
+};
+
+describePerformanceTest('getEntitiesPerformanceTest', () => {
+  const databases = TestDatabases.create({
+    ids: [/* 'MYSQL_8', */ 'POSTGRES_16', /* 'POSTGRES_12',*/ 'SQLITE_3'],
+    disableDocker: false,
+  });
+
+  it.each(databases.eachSupportedId())(
+    'fetch entities, %p',
+    async databaseId => {
+      const knex = await databases.init(databaseId);
+      await applyDatabaseMigrations(knex);
+
+      const load: SyntheticLoadOptions = {
+        baseEntitiesCount: 2000,
+        baseRelationsCount: 3,
+        baseRelationsSkew: 0.3,
+        childrenCount: 3,
+      };
+
+      traceLog('Starting test backend');
+
+      const backend = await startTestBackend({
```

Comment:
For future, would be nice to be able to profile the test backend in these tests. Don't know how to set that up; probably with node inspect flag + add some sleep before the start to have time to connect the profiler or something. Even the performance tests give some idea how things work, they really don't provide information where the bottlenecks are

Reply:
Yep indeed. I'm rewriting the test now to do all of the database setup in a beforeAll so that measurements of individual tests are reliable and don't include setup time. That will help a bit (you can easily throw in 50 variations of a request onto the single already-hot catalog etc). So it's a small step toward that at least

Reply:
![Screenshot 2024-09-14 at 12 33 19](https://github.com/user-attachments/assets/4657381b-736c-4735-914f-0796e7e8ee43)

Reply:
Nice 👌